src/Sentry.Unity/Integrations/ANRIntegration.cs

```diff
@@ -0,0 +1,134 @@
+using System;
+using System.Collections;
+using System.Diagnostics;
+using System.Threading;
+using Sentry.Extensibility;
+using Sentry.Integrations;
+using UnityEngine;
+
+namespace Sentry.Unity
+{
+    internal class ANRIntegration : ISdkIntegration
+    {
+        private static ANRWatchDog _watchdog = new ANRWatchDog();
+        private SentryMonoBehaviour _monoBehaviour;
+
+        public ANRIntegration(SentryMonoBehaviour monoBehaviour)
+        {
+            _monoBehaviour = monoBehaviour;
+        }
+
+        public void Register(IHub hub, SentryOptions options)
+        {
+            _watchdog.StartOnce(options.DiagnosticLogger, _monoBehaviour);
+            _watchdog.OnApplicationNotResponding += (_, e) => hub.CaptureException(e);
+        }
+    }
+
+    internal class ANRWatchDog
+    {
+        private int _checkIntervalMilliseconds;
+
+        // Note: we don't sleep for the check interval duration to prevent the thread from blocking shutdown.
+        // Note 2: if this is higher than _checkIntervalMilliseconds, the lower value is used.
+        private const int sleepInterval = 1000;
+
+        private CancellationTokenSource _cancel = new CancellationTokenSource();
+        private IDiagnosticLogger? _logger;
+        private SentryMonoBehaviour _monoBehaviour = null!;
+        private bool _uiIsResponsive = false;
+        private bool _reported = false; // don't report the same ANR instance multiple times
+        private Thread _thread = null!;
+        internal event EventHandler<ApplicationNotResponding> OnApplicationNotResponding = delegate { };
+
+        internal ANRWatchDog(int checkIntervalMilliseconds = 5000)
+        {
+            _checkIntervalMilliseconds = checkIntervalMilliseconds;
+        }
+
+        internal void StartOnce(IDiagnosticLogger? logger, SentryMonoBehaviour monoBehaviour)
+        {
+            // Start the thread, if not yet running.
+            lock (_cancel)
+            {
+                if (_thread is null)
+                {
+                    _monoBehaviour = monoBehaviour;
+                    _logger = logger;
+                    _thread = new Thread(Run)
+                    {
+                        Name = "Sentry-ANR-WatchDog"
+                    };
+                    _thread.Start();
+
+                    // Stop the thread when the app is being shut down.
+                    _monoBehaviour.Application.Quitting += () =>
+                    {
+                        _cancel.Cancel();
```

Comment:
Cancel can throw. Should we have `Join` in a `finally` block and log out on `catch`?

Reply:
removed the cancellation token in favour of a plain bool, which in the meantime I've found out is atomic in c#